com.unity.sg2/Editor/GraphDeltaRegistry/Registry/GraphType.cs

```diff
@@ -11,12 +12,123 @@ namespace UnityEditor.ShaderGraph.GraphDelta
+                length = truncate(length, fieldLength);
+                height = truncate(truncate(length, height), fieldHeight); // height truncates by length also.
+
+                // special case- if there is any non-matrix connection, the output dynamic ports will also be nonMatrix.
+                hasVector |= fieldHeight == 1;
+
+                // precision and primitive type always promote.
+                precision = Mathf.Max(precision, (int)GetPrecision(field));
+                primitive = Mathf.Max(primitive, (int)GetPrimitive(field));
+            }
+            return hasVector;
+        }
+
+        #endregion
+
+        public static void ResolveDynamicPorts(NodeHandler node)
+        {
+            // TODO: Resolve precision/primitive too.
+            bool hasVector = calcResolve(node, out var resolvedLength, out var resolvedHeight, out var resolvedPrecision, out var resolvedPrimitive);
+
+            // foreach graphType port.
+            foreach (var port in node.GetPorts().Where(e => e.GetTypeField().GetRegistryKey().Name == GraphType.kRegistryKey.Name))
+            {
+                var field = port.GetTypeField();
+                var precision = GetPrecision(field);
+                var primitive = GetPrimitive(field);
+                GetDim(field, out var length, out var height);
+                GetDynamic(field, out var isLenDyn, out var isHgtDyn, out var isPrecisionDyn, out var isPrimDyn);
+
+                length = isLenDyn ? resolvedLength : length;
+                height = isHgtDyn ? resolvedHeight : height;
+                precision = isPrecisionDyn ? (GraphType.Precision)resolvedPrecision : precision;
+                primitive = isPrimDyn ? (GraphType.Primitive)resolvedPrimitive : primitive;
+
+                // resolvedHeight only applies for matrices-- if we are connected to a non-matrix we need to ignore it.
+                if (isHgtDyn)
+                {
+                    var connectedField = port.IsInput ? port.GetConnectedPorts().FirstOrDefault()?.GetTypeField() : null;
+                    if (!port.IsInput && hasVector || // output port will always resolve to vector if one of the input ports does.
+                        port.IsInput && connectedField == null && hasVector || // as will disconnected input ports.
+                        port.IsInput && connectedField != null && GetHeight(connectedField) == GraphType.Height.One)
+                            height = 1;
+                }
+
+                InitGraphType(
+                    field,
+                    length: (GraphType.Length)length,
+                    height: (GraphType.Height)height,
+                    precision: precision,
+                    primitive: primitive,
+                    lengthDynamic: isLenDyn,
+                    heightDynamic: isHgtDyn,
+                    primitiveDynamic: isPrimDyn,
+                    precisionDynamic: isPrecisionDyn);
+            }
+        }
+
```

Comment:
Nothing too special happening here.

I did some crunching on SG1 type conversion, and came up with a simpler set of resolution requirements than what we had. Unfortunately, there is a bit of special casing for height.